esphome/components/atm90e26/sensor.py

```diff
@@ -0,0 +1,155 @@
+CONF_GAIN_CT = "gain_ct"
+LINE_FREQS = {
+    "50HZ": 50,
+    "60HZ": 60,
+}
+PGA_GAINS = {
+    "1X": 0x4,
+    "4X": 0x0,
+    "8X": 0x1,
+    "16X": 0x2,
+    "24X": 0x3,
+}
+
+atm90e26_ns = cg.esphome_ns.namespace("atm90e26")
+ATM90E26Component = atm90e26_ns.class_(
+    "ATM90E26Component", cg.PollingComponent, spi.SPIDevice
+)
+
+CONFIG_SCHEMA = (
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(ATM90E26Component),
+            cv.Optional(CONF_VOLTAGE): sensor.sensor_schema(
+                unit_of_measurement=UNIT_VOLT,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_VOLTAGE,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_CURRENT): sensor.sensor_schema(
+                unit_of_measurement=UNIT_AMPERE,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_CURRENT,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_POWER): sensor.sensor_schema(
+                unit_of_measurement=UNIT_WATT,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_POWER,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_REACTIVE_POWER): sensor.sensor_schema(
+                unit_of_measurement=UNIT_VOLT_AMPS_REACTIVE,
+                icon=ICON_LIGHTBULB,
+                accuracy_decimals=2,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_POWER_FACTOR): sensor.sensor_schema(
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_POWER_FACTOR,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Optional(CONF_FORWARD_ACTIVE_ENERGY): sensor.sensor_schema(
+                unit_of_measurement=UNIT_WATT_HOURS,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_ENERGY,
+                state_class=STATE_CLASS_TOTAL_INCREASING,
+            ),
+            cv.Optional(CONF_REVERSE_ACTIVE_ENERGY): sensor.sensor_schema(
+                unit_of_measurement=UNIT_WATT_HOURS,
+                accuracy_decimals=2,
+                device_class=DEVICE_CLASS_ENERGY,
+                state_class=STATE_CLASS_TOTAL_INCREASING,
+            ),
+            cv.Optional(CONF_FREQUENCY): sensor.sensor_schema(
+                unit_of_measurement=UNIT_HERTZ,
+                icon=ICON_CURRENT_AC,
+                accuracy_decimals=1,
+                state_class=STATE_CLASS_MEASUREMENT,
+            ),
+            cv.Required(CONF_LINE_FREQUENCY): cv.enum(LINE_FREQS, upper=True),
+            cv.Optional(CONF_METER_CONSTANT, default=3200.0): cv.positive_float,
+            cv.Optional(CONF_PL_CONST, default=1429876): cv.uint32_t,
+            cv.Optional(CONF_GAIN_METERING, default=7481): cv.uint16_t,
+            cv.Optional(CONF_GAIN_VOLTAGE, default=26400): cv.uint16_t,
```

Comment:
Voltage gain should also have range validation. And possibly the metering gain as well?

```suggestion
            cv.Optional(CONF_GAIN_VOLTAGE, default=26400): cv.int_range(min=0, max=32767),
```

Reply:
As far as I can see both voltage and metering gain (ugain and lgain) can take a full uint16 (up to 65k), so this should already be correct. Unless you're saying that using `uint16_t` doesn't enforce 65535 as maximum value?

Reply:
Right, all good then. You might want to clarify that in the docs though, which currently imply that 32768 is the maximum value:

> This would imply a gain_voltage value of 230 / 86.6 * 26400 = 70115. However, the chip’s application note says this value must be below 32768.

Reply:
I'm sorry, you were right. Looking with fresh eyes: ugain (volt_gain) should be limited to 15 bits (as documented) while IgainL (ct_gain) shouldn't be. I had swapped them about in the python sensor. I think I confused myself by having the datasheet naming, easier-to-read yaml naming, and then an additional internal name. I've removed the additional internal level of naming to be consistent with the yaml names to make things more obvious. Thanks for the close read.